core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,217 @@
+  }
+
+  public UpdateSnapshotReferencesOperation removeBranch(String name) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    Preconditions.checkArgument(!name.equals(SnapshotRef.MAIN_BRANCH), "Cannot remove main branch");
+    SnapshotRef ref = updatedRefs.remove(name);
+    Preconditions.checkArgument(ref != null, "No such branch with name %s", name);
+    Preconditions.checkArgument(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation removeTag(String name) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    SnapshotRef ref = updatedRefs.remove(name);
+    Preconditions.checkArgument(ref != null, "No such tag with name %s", name);
+    Preconditions.checkArgument(ref.isTag(), "Ref with name %s is a branch not a tag", name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation renameBranch(String name, String newName) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    Preconditions.checkNotNull(newName, "Branch name cannot be null");
+    Preconditions.checkArgument(!name.equals(SnapshotRef.MAIN_BRANCH), "Cannot rename main branch");
+    SnapshotRef ref = updatedRefs.get(name);
+    Preconditions.checkArgument(ref != null, "No such ref with name %s", name);
+    Preconditions.checkArgument(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    SnapshotRef existing = updatedRefs.put(newName, ref);
+    ValidationException.check(existing == null, "Ref %s already exists", newName);
+    updatedRefs.remove(name, ref);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceBranch(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = updatedRefs.get(name);
+    Preconditions.checkArgument(ref != null, "Branch %s does not exist", name);
+    Preconditions.checkArgument(ref.isBranch(), "Ref with name %s is a tag not a branch", name);
+    SnapshotRef updatedRef = SnapshotRef.builderFrom(ref, snapshotId).build();
+    updatedRefs.put(name, updatedRef);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceBranch(String name, String source) {
+    return replaceBranch(name, source, false);
+  }
+
+  public UpdateSnapshotReferencesOperation fastForward(String name, String source) {
+    return replaceBranch(name, source, true);
+  }
+
+  private UpdateSnapshotReferencesOperation replaceBranch(String name, String source, boolean fastForward) {
+    Preconditions.checkNotNull(name, "Target branch cannot be null");
+    Preconditions.checkNotNull(source, "Source ref cannot be null");
+    SnapshotRef sourceRef = updatedRefs.get(source);
+    SnapshotRef refToUpdate = updatedRefs.get(name);
+    Preconditions.checkArgument(sourceRef != null, "Ref %s does not exist", source);
+    Preconditions.checkArgument(refToUpdate != null, "Branch %s does not exist", name);
```

Comment:
I think these error messages should also distinguish between source and target branches a bit more clearly.

Reply:
Updated